v2/car_reader_seeker.go

```diff
@@ -0,0 +1,107 @@
+package car
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	"golang.org/x/xerrors"
+)
+
+// CarReaderSeeker wraps CarOffsetWriter with a ReadSeeker implementation.
+// Note that Read and Seek are not thread-safe, they must not be called
+// concurrently.
+type CarReaderSeeker struct {
+	parentCtx context.Context
+	size      uint64
+	offset    int64
+
+	cow           *CarOffsetWriter // 🐮
+	reader        *io.PipeReader
+	writeCancel   context.CancelFunc
+	writeComplete chan struct{}
+}
+
+var _ io.ReadSeeker = (*CarReaderSeeker)(nil)
+
+func NewCarReaderSeeker(ctx context.Context, payloadCid cid.Cid, bstore blockstore.Blockstore, size uint64) *CarReaderSeeker {
```

Comment:
are you confident you'll know the overall size of the car at construction time here?

Reply:
We are for the Estuary use case - if we don't know it then it makes seeking from the end infeasible

Reply:
do you need to seek from the end? the current read-seeker in this library disallows seeking from end since it doesn't know total length in advance

Reply:
http.ServeContent() gets the size of the file by [seeking to the end of it](https://github.com/golang/go/blob/a991d9dc27bda23018e23488806c8f8d027e4f7b/src/net/http/fs.go#L192-L205). I couldn't see a way to get around that without copying a whole bunch of code.

Reply:
Potentially we could do something like providing the size of the file as an option to CarReaderSeeker, and if it's not set we disallow seeking from the end?

Reply:
makes sense. I guess worth including then